### PR TITLE
mypy strict compliant

### PR DESCRIPTION
**Pull request by nathanrpage97**
## Type of changes

- [ ] Bug fix
- [ ] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [x] Other

## Checklist

- [x] I've run the latest [black](https://github.com/psf/black) with default args on new code.
- [ ] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [ ] I've added tests for new code.
- [x] I accept that @willmcgugan may be pedantic in the code review.

## Description

This is a first draft of implementing Mypy strict. I tried to minimize using `type:ignore`, `casts`, and `Any` features. Let me know what you think.

### CODE REVIEW

File: rich/__main__.py [hunk 1]
```diff
@@ -88,7 +95,7 @@ def make_test_card() -> Table:
         ),
     )
 
-    def comparison(renderable1, renderable2) -> Table:
+    def comparison(renderable1: Any, renderable2: Any) -> Table:
```
Comment (willmcgugan) [inline rich/__main__.py:98]: Should really be `RenderableType` rather than Any.

File: rich/_emoji_replace.py [hunk 2]
```diff
@@ -1,11 +1,16 @@
-from typing import Match
+from typing import Callable, Match
 
 import re
 
 from ._emoji_codes import EMOJI
 
 
-def _emoji_replace(text: str, _emoji_sub=re.compile(r"(:(\S*?):)").sub) -> str:
+_EmojiSub = Callable[[Callable[[Match[str]], str], str], str]
```
Comment (willmcgugan) [inline rich/_emoji_replace.py:8]: This was hard to parse in my head, but I double checked it and it's entirely correct. Bravo.

The thing is that readability has been thrown out for the sake of correctness, which I'm no fan of. Could we break this up a bit to make it more readable? Something along these lines:

```python
_ReStringMatch = Match[str]  # regex match object
_ReSubCallable = Callable[[_ReStringMatch], str] # Callable invoked by re.sub
_EmojiSubMethod = Callable[_ReSubCallable, str]. # Sub method of a compiled re
```

File: rich/_timer.py [hunk 3]
```diff
@@ -6,10 +6,11 @@
 from time import time
 
 import contextlib
+from typing import Iterator
 
 
 @contextlib.contextmanager
-def timer(subject: str = "time"):
+def timer(subject: str = "time") -> Iterator[None]:
```
Comment (willmcgugan) [inline rich/_timer.py:13]: I think `typing.Generator` would be more accurate here.

File: rich/_windows.py [hunk 4]
```diff
@@ -18,7 +19,10 @@ class WindowsConsoleFeatures:
     from ctypes import wintypes
     from ctypes import LibraryLoader
 
-    windll = LibraryLoader(ctypes.WinDLL)  # type: ignore
+    if sys.platform == "win32":
```
Comment (willmcgugan) [inline rich/_windows.py:22]: What was this solving

I suspect that if windll is None, it will break on L38.

File: rich/_windows.py [hunk 5]
```diff
@@ -52,8 +56,8 @@ def get_windows_console_features() -> WindowsConsoleFeatures:
         result = _GetConsoleMode(handle, console_mode)
         vt = bool(result and console_mode.value & ENABLE_VIRTUAL_TERMINAL_PROCESSING)
         truecolor = False
-        if vt:
-            win_version = sys.getwindowsversion()  # type: ignore
+        if sys.platform == "win32" and vt:
```
Comment (willmcgugan) [inline rich/_windows.py:59]: Can we get here if we are not on Windows?

File: rich/console.py [hunk 6]
```diff
@@ -294,7 +294,7 @@ def __enter__(self) -> "Capture":
         self._console.begin_capture()
         return self
 
-    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
```
Comment (willmcgugan) [inline rich/console.py:297]: I can never remember the types here, which is why omitted them but let's be accurate since we are --strict.

This is the types, copied from another project. Python2 syntax. 

``` python
    def __exit__(
        self,
        exc_type,  # type: Optional[Type[BaseException]]
        exc_value,  # type: Optional[BaseException]
        traceback,  # type: Optional[TracebackType]
    ):
```

File: rich/console.py [hunk 7]
```diff
@@ -856,7 +858,7 @@ def is_terminal(self) -> bool:
         """
         if self._force_terminal is not None:
             return self._force_terminal
-        isatty = getattr(self.file, "isatty", None)
+        isatty: Optional[Callable[..., bool]] = getattr(self.file, "isatty", None)
```
Comment (willmcgugan) [inline rich/console.py:861]: This takes no parameters, so I think the type should be this:

```
Optional[Callable[[], bool]] 
```

File: rich/containers.py [hunk 8]
```diff
@@ -82,8 +84,8 @@ def __getitem__(self, index: int) -> "Text":
     def __getitem__(self, index: slice) -> "Lines":
         ...
 
-    def __getitem__(self, index):
-        return self._lines[index]
+    def __getitem__(self, index: Union[slice, int]) -> Union["Text", "Lines"]:
```
Comment (willmcgugan) [inline rich/containers.py:87]: I think the return types should be Text or List.

File: rich/control.py [hunk 9]
```diff
@@ -157,7 +157,7 @@ def __rich_console__(
             yield self.segment
 
 
-def strip_control_codes(text: str, _translate_table=_CONTROL_TRANSLATE) -> str:
+def strip_control_codes(text: str, _translate_table: Any = _CONTROL_TRANSLATE) -> str:
```
Comment (willmcgugan) [inline rich/control.py:160]: I think that table should be `Dict[int, None]`

File: rich/markdown.py [hunk 10]
```diff
@@ -107,7 +108,7 @@ class Paragraph(TextElement):
     justify: JustifyMethod
 
     @classmethod
-    def create(cls, markdown: "Markdown", node) -> "Paragraph":
+    def create(cls, markdown: "Markdown", node: Any) -> "Paragraph":
```
Comment (willmcgugan) [inline rich/markdown.py:111]: I think node should be "MarkdownElement"

File: rich/markup.py [hunk 11]
```diff
@@ -36,7 +36,12 @@ def markup(self) -> str:
         )
 
 
-def escape(markup: str, _escape=re.compile(r"(\\*)(\[[a-z#\/].*?\])").sub) -> str:
+_EscapeSub = Callable[[Callable[[Match[str]], str], str], str]
```
Comment (willmcgugan) [inline rich/markup.py:39]: Could also use breaking up in to bite size chunks, for readability sake.

File: rich/traceback.py [hunk 12]
```diff
@@ -47,7 +47,7 @@ def install(
     word_wrap: bool = False,
     show_locals: bool = False,
     indent_guides: bool = True,
-) -> Callable:
+) -> Callable[..., Any]:
```
Comment (willmcgugan) [inline rich/traceback.py:50]: Let's type this. See L72 for the signature.

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Forgive my initial reluctance, but I think this exercise is worthwhile. Mostly requests for extra detail in the typing, and maybe a hypothetical bug in _windows.

**Comment by nathanrpage97** [comment]
Thanks for the feedback, I will work on these changes later this week.

File: rich/_timer.py (hunk 3 above)
Comment (nathanrpage97) [reply rich/_timer.py:13]: https://github.com/python/typeshed/blob/master/stdlib/contextlib.pyi#L25 shows it takes in an iterator.

File: rich/_windows.py (hunk 4 above)
Comment (nathanrpage97) [reply rich/_windows.py]: Reverting to old method with a type ignore

**Comment by willmcgugan** [comment]
HI @nathanrpage97 I've created a merge conflict for you. Would you mind fixing that? Is this ready for another review?

I'm going to publish a new version soon, and it would be nice to get this in.

**Comment by willmcgugan** [comment]
Fixed merge conflict. This looks good, thanks!
